libraries/ms-common/src/stm32f0xx/soft_timer.c

```diff
@@ -0,0 +1,173 @@
+    s_soft_timer_elems[i].inuse = false;
+  }
+
+  // Enable the interrupts. TIM2 uses IRQ channel 15.
+  stm32f0xx_interrupt_nvic_enable(15, INTERRUPT_PRIORITY_NORMAL);
+  TIM_ITConfig(TIM2, TIM_IT_Update, ENABLE);
+}
+
+StatusCode soft_timer_start(uint32_t duration_us, SoftTimerCallback callback, void *context,
+                            SoftTimerID *timer_id) {
+  // Get the time this was called.
+  uint32_t curr_time = TIM_GetCounter(TIM2);
+
+  // Enable a critical section.
+  bool critical = critical_section_start();
+
+  for (uint32_t i = 0; i < SOFT_TIMER_MAX_TIMERS; i++) {
+    if (!s_soft_timer_elems[i].inuse) {
+      // Look for an empty timer.
+
+      if (s_active_timer_id >= SOFT_TIMER_MAX_TIMERS) {
+        // New timer will be the only timer. Start it.
+        s_active_timer_id = i;
+        prv_start_timer(duration_us);
+      } else if (duration_us < s_soft_timer_elems[s_active_timer_id].duration - curr_time) {
+        // New timer will run be before the active timer. Update and start it. Use as close to now
+        // as possible for the update.
+        prv_soft_timer_update(TIM_GetCounter(TIM2));
+        s_active_timer_id = i;
+        prv_start_timer(duration_us);
+      } else {
+        // Add the current time to cancel what was on the clock when called.
+        duration_us += curr_time;
+      }
+      // Otherwise, the new timer is longest ignore it until later.
+
+      // Actually populate the timer. Do this after the previous steps as this is a critical section
+      // and this timer shouldn't be updated in prv_soft_timer_update.
+      s_soft_timer_elems[i].duration = duration_us;
+      s_soft_timer_elems[i].callback = callback;
+      s_soft_timer_elems[i].context = context;
+      s_soft_timer_elems[i].inuse = true;
+      *timer_id = i;
+      critical_section_end(critical);
+      return STATUS_CODE_OK;
+    }
+  }
+
+  // Out of timers.
+  critical_section_end(critical);
+  return status_msg(STATUS_CODE_RESOURCE_EXHAUSTED, "Out of software timers.");
+}
+
+bool soft_timer_inuse(void) {
+  for (uint32_t i = 0; i < SOFT_TIMER_MAX_TIMERS; i++) {
+    if (s_soft_timer_elems[i].inuse) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+// TIM2 Interrupt handler as defined in stm32f0xx.h
+void TIM2_IRQHandler(void) {
+  if (TIM_GetITStatus(TIM2, TIM_IT_Update) != RESET) {
+    // Stop the timer to prevent repeat triggers if it is a really short period.
+    TIM_Cmd(TIM2, DISABLE);
+
+    // Update the timer assuming the active timer expired.
+    prv_soft_timer_update(s_active_timer_duration);
+
+    // Start the next active timer if it exists as determined by the update.
+    if (s_active_timer_id < SOFT_TIMER_MAX_TIMERS) {
+      prv_start_timer(s_soft_timer_elems[s_active_timer_id].duration);
+      return;
+    }
+
+    // Clear the pending bit to stop a repeat trigger if there are no new timers.
+    TIM_ClearITPendingBit(TIM2, TIM_IT_Update);
```

Comment:
I'm pretty sure the system clock is currently running at 8Mhz. We should increase it 48Mhz.
Afaik it's due to buffered writes and clock mismatches between the system and peripheral clocks.

http://false.ekta.is/2012/06/interrupt-service-routines-double-firing-on-stm32/comment-page-1/